src/repart/repart.c

```diff
@@ -7821,6 +7872,67 @@ static int parse_partition_types(const char *p, GptPartitionType **partitions, s
         return 0;
 }
 
+static int parse_join_signature(const char *p, Set **verity_settings_map) {
+        _cleanup_(verity_settings_freep) VeritySettings *verity_settings = NULL;
+        _cleanup_free_ char *root_hash = NULL;
+        _cleanup_free_ void *content = NULL;
+        const char *signature;
+        size_t len;
+        int r;
+
+        assert(p);
+        assert(verity_settings_map);
+
+        r = extract_first_word(&p, &root_hash, ":", 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse signature parameter '%s': %m", p);
+        if (!p)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Expected hash:sig");
+        if ((signature = startswith(p, "base64:"))) {
+                r = unbase64mem(signature, &content, &len);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse root hash signature '%s': %m", signature);
+        } else {
+                r = read_full_file(p, (char**) &content, &len);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read root hash signature file '%s': %m", p);
+        }
+        if (len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Empty verity signature specified.");
+        if (len > VERITY_SIG_SIZE)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Verity signatures larger than %llu are not allowed.",
+                                       VERITY_SIG_SIZE);
+
+        verity_settings = new(VeritySettings, 1);
+        if (!verity_settings)
+                return log_oom();
+
+        *verity_settings = (VeritySettings) {
+                .root_hash_sig = TAKE_PTR(content),
+                .root_hash_sig_size = len,
+        };
+
+        r = unhexmem(root_hash, &content, &len);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse root hash '%s': %m", root_hash);
+        if (len < sizeof(sd_id128_t))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Root hash must be at least 128-bit long: %s",
+                                       root_hash);
+
+        verity_settings->root_hash = TAKE_PTR(content);
```

Comment:
We do need the `free_and_replace()` here no? Actually I'm very confused, what's the point of initializing `verity_settings` above if you just overwrite it again here?

Reply:
it's different fields? roothash vs signature